src/tensor.hh

```diff
@@ -0,0 +1,297 @@
+  int n_dims;               // number of dimensions
+
+  TensorMode() :
+    batched(false), torch_dtype(DEFAULT_TORCH_TYPE),
+    cpp_dtype(DEFAULT_CPP_TYPE), n_dims(2) {}
+
+  TensorMode(bool batched, std::string &torch_dtype,
+             std::string &cpp_dtype, int n_dims) :
+             batched(batched), torch_dtype(torch_dtype),
+             cpp_dtype(cpp_dtype), n_dims(n_dims) {}
+
+  TensorMode(const TensorMode &other) :
+    batched(other.batched), torch_dtype(other.torch_dtype),
+    cpp_dtype(other.cpp_dtype), n_dims(other.n_dims) {}
+
+  TensorMode(TensorMode &&other) :
+    batched(other.batched), torch_dtype(std::move(other.torch_dtype)),
+    cpp_dtype(std::move(other.cpp_dtype)), n_dims(other.n_dims) {}
+
+  TensorMode& operator=(const TensorMode &other) {
+    batched = other.batched;
+    torch_dtype = other.torch_dtype;
+    cpp_dtype = other.cpp_dtype;
+    n_dims = other.n_dims;
+    return *this;
+  }
+
+  static bool is_tensor(const std::string &input) {
+    return input.find("tensor") != input.npos;
+  }
+
+  /*
+   * Tensor type notation in GAP-L input<> declaration: "["batched"][ND]tensor[dtype]"
+   * allowed dtypes: F32 (default), F64, I32, I64
+   * max number of dimensions (includes batch dimension (if it exists)): 99 (default: 2)
+   * examples:
+   *  - input<batchedtensorF64> : batched 2D tensor containing float64 values
+   *  - input<batched4DtensorI32> : batched 4D tensor containing int32 values
+   *  - input<tensorF64> : non-batched 2D tensor containing float64 values
+   *  - input<batchedtensor> : batched 2D tensor containing float32 values
+   */
+  static TensorMode get_tensor_mode(const std::string &input) {
+    assert(is_tensor(input));
+
+    static hashtable<std::string, std::pair<std::string, std::string>>
+    torch_type = {{"I32", {"torch:kInt32", "int"}},
+                  {"I64", {"torch::kInt64", "long"}},
+                  {"F32", {"torch:kFloat32", "float"}},
+                  {"F64", {"torch::kFloat64", "double"}}};
+
+    // check if input Tensor is batched (default: false)
+    bool batched = input.find("batched") != input.npos;
+
+    // check the number of dims of the input Tensor (default: 2)
+    std::regex regex("[a-zA-Z0-9]*([0-9]{1,2})D[a-zA-Z0-9]*");
+    std::smatch match;
+    std::regex_search(input, match, regex);
+
+    int n_dims;
+    if (match.empty()) {
+      n_dims = 2;
+    } else {
+      try {
+        n_dims = std::stoi(match[1].str());
+      } catch (const std::exception &e) {
+        std::cerr << "Couldn't convert " << match[1].str() << " to a number.\n";
+        std::exit(1);
+      }
+    }
+
+    // check the datatype of the tensor values
+    std::string torch_dtype = DEFAULT_TORCH_TYPE;
+    std::string cpp_dtype = DEFAULT_CPP_TYPE;
+
+    for (auto& key_val : torch_type) {
+      if (input.find(key_val.first) != input.npos) {
+        torch_dtype = key_val.second.first;
+        cpp_dtype = key_val.second.second;
+        break;
+      }
+    }
+
```

Comment:
If the user misspells "tensor" in the input declaration, this will already be caught in `str_to_mode` in `input.cc`. I added another warning for the specified tensor type if the parser can't determine it.